src/lib/datamodel/ClusterBasic.h

```diff
@@ -0,0 +1,71 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file contains definitions for CHIP Cluster Basic
+ *
+ */
+
+#ifndef CHIPCLUSTERBASIC_H_
+#define CHIPCLUSTERBASIC_H_
+
+#include <datamodel/Cluster.h>
+
+namespace chip {
+namespace DataModel {
+
+/* Cluster ID */
+static const uint16_t kClusterIdBase = 0x0000;
+
+/* Attribute IDs */
+static const uint16_t kAttributeIdZCLVersion         = 0x0000;
+static const uint16_t kAttributeIdApplicationVersion = 0x0001;
+static const uint16_t kAttributeIdStackVersion       = 0x0002;
+static const uint16_t kAttributeIdHWVersion          = 0x0003;
+
+/**
+ * @brief
+ *   This class implements the Base cluster as defined in the CHIP specification.
+ */
+class ClusterBasic : public Cluster
```

Comment:
The class hierarchy designed here (ClusterBasic -> Cluster, which contains Attributes -> Values) is I think much too OO-ified to be performant on embedded targets.

There should be a separation between the specification for a cluster (the set of attributes it has, the type of each attribute, their ranges, etc), from the actual data backing that cluster. The former is typically const type information, since the definition for what a cluster is is immutable, lending itself as well to code-generation. The actual dynamic data backing each attribute is controlled by the application however, and is in RAM.

Here in this CL, the application developer is forced to define clusters by defining themselves what composes that cluster (making it error-prone), as well as placing a lot of pressure on RAM.

Reply:
> Left a bunch of comments around the approach in the CL.
> 
> Overall, while I like the direction of application-centric ownership of data, I'm very concerned about the intermix of const and non-const data, as well as the memory cost of the architecture. It's fairly heavy, and I don't think will scale well to more complicated data types.

const can be pulled out with cluster/attribute/value-specific classes, basically putting them in code?

Reply:
Not sure I follow - could you provide an example?

Reply:
e.g. for clusterid, instead of making a member variable
```
class Cluster {
...
    virtual Id() const = 0;
};
class LevelControl : public Cluster {
     static const kId = 0x0008; // from spec
     
     virtual Id() const { return kId; }
     ... 
};

Reply:
we'd have to be more clever with the structure of the attributes

Reply:
Could also be done with templates, so replaces the vtable overhead with `.text`

```c++
typedef uint16_t ClusterId_t;
template <ClusterId_t tClusterId, uint16_t tAttributeNo>
class Cluster
{
public:
    static const ClusterId_t mClusterId = tClusterId;
    Attribute mAttributes[tAttributeNo];
};


class LevelControl : public Cluster<0x0008, 4>
{
public:
};
```

Reply:
That approach starts to break down fairly quickly for other const information - e.g the list of attributes in a given cluster. This is represented by the`mAttrs` private member in the Cluster class, but that clearly cannot be made const.

The only scalable approach is to actually separately represent the schema-derived information as either code, or better yet, tabular data that is then passed in as an argument to a concrete, mutable, non-const cluster data object.

This is the approach already in place in the existing Silabs code-base - why are we altering it?

Reply:
@rwalker-apple I think your example will not actually save any code -- last time I ran into examples like the above there needed to be a `final` keyword on the `LevelControl::Id` method for the compiler to actually collapse and inline the OO structures. Without the `final` I believe in most invocations a call to `Id` will actually involve a virtual call, etc (there may be some cases where compiler can prove to itself that `LevelControl` is only a `LevelControl` and not a subclass, but those are incredibly infrequent).   And if you are able to specify `final`, it often means that the class hierarchy is not as deep or does not provide as much utility as one had expected. 

This pattern looks great as a design principle; in products we've shipped, we have had better success with other structures (e.g. along the lines of what @mrjerryjohns was getting at).  And I have spent more time time than I care to admit ripping out interfaces / pure virtual classes from a codebase to squeeze the code into available space.

Reply:
> That approach starts to break down fairly quickly for other const information - e.g the list of attributes in a given cluster. This is represented by the`mAttrs` private member in the Cluster class, but that clearly cannot be made const.
> 
> The only scalable approach IMHO is to actually separately represent the schema-derived information as either code, or better yet, tabular data that is then passed in as an argument to a concrete, mutable, non-const cluster data object.
> 
> This is the approach already in place in the existing Silabs code-base - why are we altering it?

In my view, this experiment is aimed at making the SiLabs ZCL implementations easier to compose and unit test and to be able to provide the implementations to customers as libraries instead of source.  The aim is more runtime conditionals/composition, less compile-time conditionals/composition.